GenericList/string_list.h

```diff
@@ -1,31 +1,78 @@
 #pragma once
 
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <memory>
+#include <cassert>
+
 class List_String {
+    //note: elements are value copies of original objects (copy-by-value, not copy-by-reference)
 public:
     List_String();
-    ~List_String();
 
-    int Capacity();
-    bool Capacity(int new_capacity);
-    int Count();
+    friend void swap(List_String& first, List_String& second);
+
+    //Rule of 3
+    ~List_String(); //Destructor
+    List_String(const List_String& other); //Copy Constructor
+    List_String& operator=(const List_String& other); //Copy assignement
+
+    //Rule of 5
+    //R-value references (&&) explained: http://thbecker.net/articles/rvalue_references/section_01.html 
+    //basic explanation: if a function argument is &&, whatever it references will stop existing at the end of the function
+    List_String(List_String&& other);
+    List_String& operator=(List_String&& other);
+
+
+
+
+    int Capacity() const;
+    int Count() const;
+
+    //Sets the capacity of the internal array to new_capacity. If new_capacity is smaller than Count, do nothing.
+    void Capacity(int new_capacity);
+
+    std::string ToString() const;
 
-    std::string ToString();
+    void Add(const std::string& new_val);
+    void Add(std::string&& new_val);
```

Comment:
This is not what I had in mind. Perfect forwarding is something that is AFAIK only used in function templates:
```suggestion
    template<typename... Args>
    void Add(Args&&... args) {
        if (capacity == count) Capacity(2 * capacity);
        new (data + count++) std::string(std::forward<Args>(args)...);
    }
```


Also, aside from move/copy constructors and assignments, if you have functions where you want support both move and copy, just pass by value.
```
void Add(std::string new_val);
```
When the caller invokes the function they can then choose to either copy and move the data like so:
```
std::string str;
list.Add(str); // copies
list.Add(std::move(str)); // moves
```

Reply:
Do you mean that `f(std::string new_val)` just copies by value regardless of argument type (lvalue-copy/rvalue-move)? In that way, how does that differ from `f(const std::string& new_val)`? Shouldn't the latter be more efficient since it produces one less intermediate copy? Is it because, when `f` requires a copy of the argument within the function, it's better to let the compiler create that copy?

In the case of `Add`, using a function template for the argument is better since it maintains the information of the value being passed (lvalue/rvalue) to pass through using std::forward, avoiding the intermediate copy(ies) `Add(T new_val)` would create - is that right?

Reply:
`f(std::string new_val)` will copy a value if you pass it a lvalue, move if you pass it an rvalue; it basically constructs the argument from your input.

You can define `f(const std::string& new_val)` AND `f(std::string&& new_val)` or just define both cases with just `f(std::string new_val)`: 
```
// case A:
void set(std::string const& val) { str = val; }
void set(std::string&& val) { str = std::move(val); }

// case B:
void set(std::string val) { str = std::move(val); }
```
The difference between them is similar to using `swap`, there's an intermediate variable created in case B. Again its cheap in most cases, but can be expensive if the class is large or doesn't benefit much from move semantics.

When designing public APIs I tend to prefer case B, less functions for users to read, less suggestions pop up in the IDE and less for you to document.

Reply:
>In the case of Add, using a function template for the argument is better since it maintains the information of the value being passed (lvalue/rvalue) to pass through using std::forward, avoiding the intermediate copy(ies) Add(T new_val) would create - is that right?

That's right. But additionally, the function template is also variadic, so it allows the `Add` function to construct the element from the arguments. For example, if you have something like:
```
struct vector3f {
    float x,y,z;
    vector3f(float x, float y, float z) : x(x), y(y), z(z) {}
};
```

You can call `Add(0.1f,0.1f,0.1f)`, and the function will construct the element directly in the memory location without an intermediate value.